src/core/vector/qgsvectorlayerfeatureiterator.cpp

```diff
@@ -1045,11 +1049,13 @@ void QgsVectorLayerFeatureIterator::FetchJoinInfo::addJoinedAttributesCached( Qg
 
 void QgsVectorLayerFeatureIterator::FetchJoinInfo::addJoinedAttributesDirect( QgsFeature &f, const QVariant &joinValue ) const
 {
+#if 0 // this is not thread safe -- we cannot access the layer here as this will be called from non-main threads.
```

Comment:
Ping @elpaso -- I've had to disable this shortcut -- there's no easy way to make this thread safe. The hasFeatures method would need to be added to QgsVectorLayerFeatureSource and all the various provider's QgsAbstractFeatureSource subclasses, but that's not a trivial task to do.